lib/portlayer/exec/base.go

```diff
@@ -198,40 +198,39 @@ func (c *containerBase) stop(ctx context.Context, waitTime *int32) error {
 
 	log.Warnf("stopping %s via hard power off due to: %s", c.ExecConfig.ID, err)
 
-	_, err = c.vm.WaitForResult(ctx, func(ctx context.Context) (tasks.Task, error) {
-		return c.vm.PowerOff(ctx)
-	})
-
-	if err != nil {
+	return c.poweroff(ctx)
+}
 
-		// It is possible the VM has finally shutdown in between, ignore the error in that case
-		if terr, ok := err.(task.Error); ok {
-			switch terr := terr.Fault().(type) {
-			case *types.InvalidPowerState:
-				if terr.ExistingState == types.VirtualMachinePowerStatePoweredOff {
-					log.Warnf("power off %s task skipped (state was already %s)", c.ExecConfig.ID, terr.ExistingState)
-					return nil
-				}
-				log.Warnf("invalid power state during power off: %s", terr.ExistingState)
+func (c *containerBase) kill(ctx context.Context) error {
+	// make sure we have vm
+	if c.vm == nil {
+		return NotYetExistError{c.ExecConfig.ID}
+	}
 
-			case *types.GenericVmConfigFault:
+	wait := 10 * time.Second // default
+	sig := string(ssh.SIGKILL)
+	log.Infof("sending kill -%s %s", sig, c.ExecConfig.ID)
 
-				// Check if the poweroff task was canceled due to a concurrent guest shutdown
-				if len(terr.FaultMessage) > 0 && terr.FaultMessage[0].Key == vmNotSuspendedKey {
-					log.Infof("power off %s task skipped due to guest shutdown", c.ExecConfig.ID)
-					return nil
-				}
-				log.Warnf("generic vm config fault during power off: %#v", terr)
+	err := c.startGuestProgram(ctx, "kill", sig)
```

Comment:
This is not correct. You are sending `kill SIGKILL`; should be sending `kill -SIGKILL`.

Why not just put the code in the for loop in shutdown() into its own function and have both places call that?